[KAFKA-38] Read records implementation

paulgrape opened the pull request:
"Signed-off-by: [Pavel Vinogradov] [vin.pavel13@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

paulgrape committed:
Read records implementation

"Signed-off-by: [Pavel Vinogradov] [vin.pavel13@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

paulgrape committed:
Minor fixes

"Signed-off-by: [Pavel Vinogradov] [vin.pavel13@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

paulgrape commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -73,7 +135,24 @@ private void submitRecordOperation(DataOperation op, OpType opType) {
+          consumerCreateFuncCache.computeIfAbsent(consConfig, ConsumerCreateFunctionImpl::new);
+      val kafkaConsumer = consumerCache.computeIfAbsent(nodeAddr, consumerConfig);
+
+      val result = kafkaConsumer.poll(Duration.ofSeconds(10));
```
How should I process received records?

akurilov replied:
happy path:
1. extract the record from the result
2. extract the record payload (value) size
3. call `countBytesDone` method for the operation instance
4. set the `SUCC` status for the operations instance

paulgrape committed:
Read records implementation (read only 1 record)

"Signed-off-by: [Pavel Vinogradov] [vin.pavel13@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

akurilov reviewed (changes requested):
Please resolve the merge conflicts w/ master

paulgrape committed:
Read records impl

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -73,7 +135,24 @@ private void submitRecordOperation(DataOperation op, OpType opType) {
+          consumerCreateFuncCache.computeIfAbsent(consConfig, ConsumerCreateFunctionImpl::new);
+      val kafkaConsumer = consumerCache.computeIfAbsent(nodeAddr, consumerConfig);
+
+      val result = kafkaConsumer.poll(Duration.ofSeconds(10));
```
ths timeout should be a configuration option, e.g.
`storage-driver-read-timeoutMillis`

paulgrape replied:
Done

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -108,6 +187,33 @@ private void submitNoop(final O op) {
     completeOperation(op, SUCC);
   }
 
+  private Properties createConfig(String nodeAddr) {
```
please make such methods package-private to make the code more testable

paulgrape replied:
Done

paulgrape committed:
Read records implementation was fixed

"Signed-off-by: [Pavel Vinogradov] [vin.pavel13@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

paulgrape committed:
Read records implementation was fixed

paulgrape committed:
gitlab-ci.yaml was fixed

paulgrape committed:
Minor fixes

"Signed-off-by: [Pavel Vinogradov] [vin.pavel13@gmail.com]"

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or

(b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or

(c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

(d) I understand and agree that this project and the contribution are public and that a record of the contribution (including all personal information I submit with it, including my sign-off) is maintained indefinitely and may be redistributed consistent with this project or the open source license(s) involved.

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -124,7 +135,31 @@ private void submitRecordOperation(DataOperation op, OpType opType, String nodeA
+      val config = configCache.computeIfAbsent(nodeAddr, this::createConfig);
+      val adminConfig =
+          adminClientCreateFuncCache.computeIfAbsent(config, AdminClientCreateFunctionImpl::new);
+      val adminClient = adminClientCache.computeIfAbsent(nodeAddr, adminConfig);
```
unused

paulgrape replied:
Removed

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -124,7 +135,31 @@ private void submitRecordOperation(DataOperation op, OpType opType, String nodeA
+          adminClientCreateFuncCache.computeIfAbsent(config, AdminClientCreateFunctionImpl::new);
+      val adminClient = adminClientCache.computeIfAbsent(nodeAddr, adminConfig);
+
+      val consConfig = consumerConfigCache.computeIfAbsent(nodeAddr, this::createConsumerConfig);
```
consumerConfig

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -124,7 +135,31 @@ private void submitRecordOperation(DataOperation op, OpType opType, String nodeA
+      val adminClient = adminClientCache.computeIfAbsent(nodeAddr, adminConfig);
+
+      val consConfig = consumerConfigCache.computeIfAbsent(nodeAddr, this::createConsumerConfig);
+      val consumerConfig =
```
consumerCreateFunc

akurilov commented on src/main/java/com/emc/mongoose/storage/driver/kafka/KafkaStorageDriver.java:
```diff
@@ -124,7 +135,31 @@ private void submitRecordOperation(DataOperation op, OpType opType, String nodeA
+          consumerCreateFuncCache.computeIfAbsent(consConfig, ConsumerCreateFunctionImpl::new);
+      val kafkaConsumer = consumerCache.computeIfAbsent(nodeAddr, consumerConfig);
+
+      val result = kafkaConsumer.poll(Duration.ofMillis(readTimeout));
```
please invoke the `startRequest`, `finishRequest` methods before this and
`startResponse`, `finishResponse` after

paulgrape replied:
`finishRequest`, `startResponse` and `finishResponse` are called in `completeOperation` method

akurilov replied:
ok, let's change this later

paulgrape committed:
Merge branch 'master' into kafka-38-read-records-implementation
